node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -205,16 +235,102 @@ impl OrderingProvider {
 			self.included_candidates.remove(&finalized_candidate);
 		}
 	}
+
+	/// Returns ancestors of `head` in the descending order, stopping
+	/// either at the block present in cache or the latest finalized block.
+	///
+	/// Suited specifically for querying non-finalized chains, thus
+	/// doesn't rely on block numbers.
+	///
+	/// Both `head` and last are **not** included in the result.
+	async fn get_block_ancestors<Sender: SubsystemSender>(
+		&mut self,
+		sender: &mut Sender,
+		mut head: Hash,
+		mut head_number: BlockNumber,
+	) -> Result<Vec<Hash>> {
+		const ANCESTRY_STEP: usize = 10;
+		const ANCESTRY_SIZE_LIMIT: usize = 1000;
+
+		let mut ancestors = Vec::new();
+
+		if self.last_observed_blocks.get(&head).is_some() {
+			return Ok(ancestors)
+		}
+
+		let finalized_block_number = get_finalized_block_number(sender).await?;
+
+		loop {
+			let (tx, rx) = oneshot::channel();
+			let hashes = {
+				sender
+					.send_message(
+						ChainApiMessage::Ancestors {
+							hash: head,
+							k: ANCESTRY_STEP,
+							response_channel: tx,
+						}
+						.into(),
+					)
+					.await;
+
+				rx.await.map_err(NonFatal::Oneshot)?.map_err(NonFatal::ChainApi)?
+			};
+
+			let earliest_block_number = head_number - hashes.len() as u32;
+			// The reversed order is parent, grandparent, etc. excluding the head.
+			let block_numbers = (earliest_block_number..head_number).rev();
+
+			for (block_number, hash) in block_numbers.zip(&hashes) {
+				// Return if we either met finalized/cached block or
+				// hit the size limit for the returned ancestry of head.
+				if self.last_observed_blocks.get(&head).is_some() ||
+					block_number <= finalized_block_number ||
+					ancestors.len() >= ANCESTRY_SIZE_LIMIT
+				{
+					return Ok(ancestors)
+				}
+
+				ancestors.push(*hash);
+			}
+
+			match hashes.last() {
+				Some(last_hash) => {
+					head = *last_hash;
+					head_number = earliest_block_number;
+				},
+				None => break,
+			}
+		}
+		return Ok(ancestors)
+	}
+}
+
+async fn send_message_fatal<Sender, Response>(
+	sender: &mut Sender,
+	message: ChainApiMessage,
+	receiver: oneshot::Receiver<std::result::Result<Response, ChainApiError>>,
+) -> FatalResult<Response>
+where
+	Sender: SubsystemSender,
+{
+	sender.send_message(message.into()).await;
+
+	receiver
+		.await
+		.map_err(|_| Fatal::ChainApiSenderDropped)?
+		.map_err(Fatal::ChainApi)
 }
 
 async fn get_block_number(
```

Comment:
dead code.

Reply:
Do you mean `get_block_number`? It's used for getting candidate comparator.

Reply:
haha, right - only looked at the diff.